lib/uplink/ex_restrict_test.go

```diff
@@ -0,0 +1,147 @@
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/lib/uplink"
+	"storj.io/storj/pkg/macaroon"
+)
+
+func RestrictAccessExampleByAdmin(ctx context.Context, satelliteAddress, apiKey string, adminEncCtx string, cfg *uplink.Config, out io.Writer) (serializedUserAPIKey string, serializedEncCtx string, err error) {
+	// Parse the API key. API keys are "macaroons" that allow you to create new, restricted API keys.
+	key, err := uplink.ParseAPIKey(apiKey)
+	if err != nil {
+		return "", "", err
+	}
+
+	// Restrict the API key to be read only and to be for just the prod and staging buckets
+	// for the path webserver/logs/
+	userAPIKey, err := key.Restrict(macaroon.Caveat{
+		DisallowWrites:  true,
+		DisallowDeletes: true,
+	})
+	if err != nil {
+		return "", "", err
+	}
+
+	// Load the existing encryption context
+	encCtx, err := uplink.ParseEncryptionCtx(adminEncCtx)
+	if err != nil {
+		return "", "", err
+	}
+
+	// Restrict the encryption context to just the prod and staging buckets
+	// for webserver/logs/
+	userAPIKey, userEncCtx, err := encCtx.Restrict(userAPIKey,
+		uplink.EncryptionRestriction{Bucket: "prod", PathPrefix: "webserver/logs"},
+		uplink.EncryptionRestriction{Bucket: "staging", PathPrefix: "webserver/logs"},
+	)
+	if err != nil {
+		return "", "", err
+	}
+
+	// Serialize the encryption context
+	serializedUserEncCtx, err := userEncCtx.Serialize()
+	if err != nil {
+		return "", "", err
+	}
+
+	fmt.Fprintln(out, "success!")
+	return userAPIKey.Serialize(), serializedUserEncCtx, nil
+}
+
+func RestrictAccessExampleByUser(ctx context.Context, satelliteAddress, apiKey string, serializedEncCtx string, cfg *uplink.Config, out io.Writer) (err error) {
+	errCatch := func(fn func() error) { err = errs.Combine(err, fn()) }
+
+	// First, create an Uplink handle.
+	ul, err := uplink.NewUplink(ctx, cfg)
+	if err != nil {
+		return err
+	}
+	defer errCatch(ul.Close)
+
+	// Parse the API key.
+	key, err := uplink.ParseAPIKey(apiKey)
+	if err != nil {
+		return err
+	}
+
+	// Open the project in question. Projects are identified by a specific Satellite and API key
+	p, err := ul.OpenProject(ctx, satelliteAddress, key)
+	if err != nil {
+		return err
+	}
+	defer errCatch(p.Close)
+
+	// Parse the encryption context
+	encCtx, err := uplink.ParseEncryptionCtx(serializedEncCtx)
+	if err != nil {
+		return err
+	}
+
+	// Open bucket
+	bucket, err := p.OpenBucket(ctx, "prod", encCtx)
+	if err != nil {
+		return err
+	}
+	defer errCatch(bucket.Close)
+
+	// Open file
+	obj, err := bucket.OpenObject(ctx, "webserver/logs/log.txt")
+	if err != nil {
+		return err
+	}
+	defer errCatch(obj.Close)
+
+	// Get a reader for the entire file
+	r, err := obj.DownloadRange(ctx, 0, -1)
+	if err != nil {
+		return err
+	}
+	defer errCatch(r.Close)
+
+	// Read the file
+	data, err := ioutil.ReadAll(r)
+	if err != nil {
+		return err
+	}
+
+	// Print it!
+	fmt.Fprintln(out, string(data))
+	return nil
+}
+
+func Example_restrictAccess() {
```

Comment:
Let's use testplanet here instead and ensure that this runs as a test.

Reply:
we want this as documentation. testplanet does run this though. see examples_test.go